go/pkg/tool/tool.go

```diff
@@ -99,18 +101,43 @@ func (c *Client) prepCommand(ctx context.Context, client *rexec.Client, actionDi
 	if _, err := c.GrpcClient.ReadProto(ctx, cmdDg, commandProto); err != nil {
 		return nil, err
 	}
-	if inputRoot == "" {
+	var inputRoot string
+	var nodeProperties map[string]*repb.NodeProperties
+	if actionRoot == "" {
 		curTime := time.Now().Format(time.RFC3339)
-		inputRoot = filepath.Join(os.TempDir(), acDg.Hash+"_"+curTime)
+		actionRoot = filepath.Join(os.TempDir(), acDg.Hash+"_"+curTime)
+		inputRoot = filepath.Join(actionRoot, "input")
 		dg, err := digest.NewFromProto(actionProto.GetInputRootDigest())
 		if err != nil {
 			return nil, err
 		}
 		log.Infof("Fetching input tree from input root digest %s into %s", dg, inputRoot)
-		_, _, err = c.GrpcClient.DownloadDirectory(ctx, dg, inputRoot, client.FileMetadataCache)
+		ts, _, err := c.GrpcClient.DownloadDirectory(ctx, dg, inputRoot, client.FileMetadataCache)
 		if err != nil {
 			return nil, err
 		}
+		nodeProperties = make(map[string]*repb.NodeProperties)
+		for path, t := range ts {
+			if t.NodeProperties != nil {
+				nodeProperties[path] = t.NodeProperties
+			}
+		}
+	} else {
+		inputRoot = filepath.Join(actionRoot, "input")
+		npPath := filepath.Join(actionRoot, "input_node_properties.textproto")
+		if _, err := os.Stat(npPath); err == nil {
```

Comment:
I tried to rewrite this with early return, but failed -- only way I saw to return early is to duplicate one of the if clauses, which was, imo, worse.